p2p/src/network/rpc/p2p_network_rpc_reducer.rs

```diff
@@ -1,33 +1,83 @@
+use std::sync::Arc;
+
 use binprot::BinProtRead;
-use mina_p2p_messages::rpc_kernel::{MessageHeader, QueryHeader, RpcMethod};
+use mina_p2p_messages::{
+    rpc,
+    rpc_kernel::{
+        MessageHeader, PayloadBinprotReader as _, QueryHeader, ResponseHeader, RpcMethod,
+        RpcQueryReadError, RpcResponseReadError,
+    },
+    v2,
+    versioned::Ver,
+};
+use openmina_core::{error, fuzz_maybe, fuzzed_maybe, Substate};
+use redux::Dispatcher;
 
-use crate::{Limit, P2pLimits};
+use crate::{
+    channels::rpc::{
+        BestTipWithProof, P2pChannelsRpcAction, P2pRpcRequest, P2pRpcResponse,
+        StagedLedgerAuxAndPendingCoinbases,
+    },
+    connection::outgoing::P2pConnectionOutgoingInitOpts,
+    disconnection::{P2pDisconnectionAction, P2pDisconnectionReason},
+    Data, Limit, P2pLimits, P2pNetworkState, P2pNetworkYamuxAction, PeerId,
+};
 
 use self::p2p_network_rpc_state::P2pNetworkRpcError;
 
 use super::*;
 
 impl P2pNetworkRpcState {
-    pub fn reducer(
-        &mut self,
+    pub fn reducer<State, Action>(
+        mut state_context: Substate<Action, State, P2pNetworkState>,
```

Comment:
@0xMimir when `Substate` is not over `Self` please add a comment mentioning what other substate is accessed, just for future reference once we figure out what is the best way to narrow it.

Reply:
looking at the rest of the code I am still not sure why this cannot be done over `Self` btw, is that really required?

Reply:
ok I see, this is what you mentioned on Slack right? regarding the selection of the substate based on the kind of action (which here happens when calling `find_rpc_state_mut`)

Reply:
But even in that case.... I think you can still use `Substate<...., Self>` ? it is just that you will have to select the substate in the caller